ecscollector/collector.go

```diff
@@ -0,0 +1,231 @@
+		"Total CPU usage.",
+		cpuLabels, nil)
+
+	cpuSystemDesc = prometheus.NewDesc(
+		"ecs_cpu_seconds_system",
+		"Total system CPU usage.",
+		labels, nil)
+
+	cpuOnlineDesc = prometheus.NewDesc(
+		"ecs_cpu_online",
+		"Number of online CPUs.",
+		labels, nil)
+
+	memUsageDesc = prometheus.NewDesc(
+		"ecs_memory_bytes",
+		"Memory usage in bytes.",
+		labels, nil)
+
+	memMaxUsageDesc = prometheus.NewDesc(
+		"ecs_memory_max_bytes",
+		"Maximum memory usage in bytes.",
+		labels, nil)
+
+	memLimitDesc = prometheus.NewDesc(
+		"ecs_memory_limit_bytes",
+		"Memory limit in bytes.",
+		labels, nil)
+
+	networkRxBytesDesc = prometheus.NewDesc(
+		"ecs_network_receive_bytes",
+		"Network recieved in bytes.",
+		networkLabels, nil)
+
+	networkRxPacketsDesc = prometheus.NewDesc(
+		"ecs_network_receive_packets",
+		"Network packets recieved.",
+		networkLabels, nil)
+
+	networkRxDroppedDesc = prometheus.NewDesc(
+		"ecs_network_receive_dropped",
+		"Network packets dropped in recieving.",
+		networkLabels, nil)
+
+	networkRxErrorsDesc = prometheus.NewDesc(
+		"ecs_network_receive_errors",
+		"Network errors in recieving.",
+		networkLabels, nil)
+
+	networkTxBytesDesc = prometheus.NewDesc(
+		"ecs_network_transmit_bytes",
+		"Network transmitted in bytes.",
+		networkLabels, nil)
+
+	networkTxPacketsDesc = prometheus.NewDesc(
+		"ecs_network_transmit_packets",
+		"Network packets transmitted.",
+		networkLabels, nil)
+
+	networkTxDroppedDesc = prometheus.NewDesc(
+		"ecs_network_transmit_dropped",
+		"Network packets dropped in transmit.",
+		networkLabels, nil)
+
+	networkTxErrorsDesc = prometheus.NewDesc(
+		"ecs_network_transmit_errors",
+		"Network errors in transmit.",
+		networkLabels, nil)
+)
+
+var labels = []string{
+	"container",
+}
+
+var cpuLabels = append(
+	labels,
+	"mode", // user, kernel, system
+)
+
+var networkLabels = append(
+	labels,
+	"device",
+)
+
+// NewCollector returns a new Collector that queries ECS metadata server
+// for ECS task and container metrics.
+func NewCollector(client *ecsmetadata.Client) prometheus.Collector {
+	return &collector{client: client}
+}
+
+type collector struct {
+	client *ecsmetadata.Client
+}
+
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- numProcsDesc
+	ch <- cpuTotalDesc
+	ch <- cpuSystemDesc
+	ch <- cpuOnlineDesc
+	ch <- memUsageDesc
+	ch <- memMaxUsageDesc
+	ch <- memLimitDesc
+	ch <- networkRxBytesDesc
+	ch <- networkRxPacketsDesc
+	ch <- networkRxDroppedDesc
+	ch <- networkRxErrorsDesc
+	ch <- networkTxBytesDesc
+	ch <- networkTxPacketsDesc
+	ch <- networkTxDroppedDesc
+	ch <- networkTxErrorsDesc
+}
+
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	ctx := context.Background()
+	metadata, err := c.client.RetrieveTaskMetadata(ctx)
+	if err != nil {
+		log.Printf("Failed to retrieve metadata: %v", err)
+		return
+	}
+	stats, err := c.client.RetrieveTaskStats(ctx)
+	if err != nil {
+		log.Printf("Failed to retrieve container stats: %v", err)
```

Comment:
should we have some kind of "up" metric to let the user know that the scrape was not successful?

Reply:
Actually if it is not up, we should return an HTTP 500 with NewInvalidMetric

https://www.robustperception.io/failing-a-scrape-with-the-prometheus-go-client

Reply:
@roidelapluie  nice, TIL!